src/merge.ts

```diff
@@ -0,0 +1,44 @@
+import { EventPayload } from './EventPayload';
+
+/**
+ * Used for merging request JSON into an existing EventPayload.
+ * Returns a new EventPayload, with the original object unaffected.
+ * The merge occurs using the following conventions:
+ *
+ *  Merging an existing key with null/undefined deletes the key.
+ *  Merging an existing key with non-null value updates the value.
+ *  Merging a non-existing key adds the key/value.
+ *  When the value is an object, the above is applied iteratively.
+ */
+export const merge = (
+  original: EventPayload,
+  newValues: Partial<Record<keyof EventPayload, unknown>>
+): EventPayload => {
+  const copy: Record<string, unknown> = { ...original };
+  const stack: Array<[Record<string, unknown>,
+     Partial<Record<keyof EventPayload, unknown>>]> = [[copy, newValues]];
+
+  while (stack.length > 0) {
+    const [target, source] = stack.pop() || [];
```

Comment:
Do you need to do `stack.pop() || []`? There should always be a value in the stack as long as the loop is running.

Reply:
While that is true, typescript still complains about it without

Type '[Record<string, unknown>, Partial<Record<keyof EventPayload, unknown>>] | undefined' is not an array type.
when I do const [target, source] = stack.pop();

I also tried const [target, source] = stack.pop()!;
but that has a warning 
Forbidden non-null assertion.eslint[@typescript-eslint/no-non-null-assertion](https://github.com/typescript-eslint/typescript-eslint/blob/v4.33.0/packages/eslint-plugin/docs/rules/no-non-null-assertion.md)

Reply:
Then in that case let's use ?? instead of ||